Mathlib/Data/List/EditDistance/Defs.lean

```diff
@@ -0,0 +1,288 @@
+  rfl
+
+theorem suffixLevenshtein_cons₁_aux {x y : {r : List δ // 0 < r.length}}
+    (w₀ : x.1[0]'x.2 = y.1[0]'y.2) (w : x.1.tail = y.1.tail) : x = y := by
+  match x, y with
+  | ⟨hx :: tx, _⟩, ⟨hy :: ty, _⟩ => simp_all
+
+theorem suffixLevenshtein_cons₁
+    (x : α) (xs ys) :
+    suffixLevenshtein C (x :: xs) ys =
+      ⟨levenshtein C (x :: xs) ys ::
+        (suffixLevenshtein C xs ys).1, by simp⟩ := by
+  induction ys
+  · case nil =>
+    dsimp [levenshtein, suffixLevenshtein]
+  · case cons y ys ih =>
+    apply suffixLevenshtein_cons₁_aux
+    · rfl
+    · rw [suffixLevenshtein_cons₂ (x :: xs), ih, impl_cons]
+      · rfl
+      · simp [suffixLevenshtein_length]
+
+theorem suffixLevenshtein_cons₁_fst (x : α) (xs ys) :
+    (suffixLevenshtein C (x :: xs) ys).1 =
+      levenshtein C (x :: xs) ys ::
+        (suffixLevenshtein C xs ys).1 := by
+  simp [suffixLevenshtein_cons₁]
+
+theorem suffixLevenshtein_cons_cons_fst_get_zero
+    (x : α) (xs y ys) (w) :
+    (suffixLevenshtein C (x :: xs) (y :: ys)).1[0] =
+      let ⟨dx, _⟩ := suffixLevenshtein C xs (y :: ys)
+      let ⟨dy, _⟩ := suffixLevenshtein C (x :: xs) ys
+      let ⟨dxy, _⟩ := suffixLevenshtein C xs ys
+      min
+        (C.delete x + dx[0])
+        (min
+          (C.insert y + dy[0])
+          (C.substitute x y + dxy[0])) := by
+  conv =>
+    lhs
+    dsimp only [suffixLevenshtein_cons₂]
+  simp only [suffixLevenshtein_cons₁]
+  rw [impl_cons_fst_zero]
+  rfl
+
+theorem suffixLevenshtein_eq_tails_map (xs ys) :
+    (suffixLevenshtein C xs ys).1 = xs.tails.map fun xs' => levenshtein C xs' ys := by
+  induction xs
+  · case nil =>
+    simp only [List.map, suffixLevenshtein_nil']
+  · case cons x xs ih =>
+    simp only [List.map, suffixLevenshtein_cons₁, ih]
+
+@[simp]
+theorem levenshtein_nil_nil : levenshtein C [] [] = 0 := by
+  simp [levenshtein]
+
+@[simp]
+theorem levenshtein_nil_cons (y) (ys) :
+    levenshtein C [] (y :: ys) = C.insert y + levenshtein C [] ys := by
+  dsimp [levenshtein]
+  congr
+  rw [List.getLast_eq_get]
+  congr
+  rw [show (List.length _) = 1 from _]
+  induction ys with
+  | nil => simp
+  | cons y ys ih =>
+    simp only [List.foldr]
+    rw [impl_length] <;> simp [ih]
+
+@[simp]
+theorem levenshtein_cons_nil (x : α) (xs : List α) :
+    levenshtein C (x :: xs) [] = C.delete x + levenshtein C xs [] :=
+  rfl
+
+@[simp]
+theorem levenshtein_cons_cons
+    (x : α) (xs : List α) (y : β) (ys : List β) :
+    levenshtein C (x :: xs) (y :: ys) =
+      min (C.delete x + levenshtein C xs (y :: ys))
+        (min (C.insert y + levenshtein C (x :: xs) ys)
+          (C.substitute x y + levenshtein C xs ys)) :=
+  suffixLevenshtein_cons_cons_fst_get_zero _ _ _ _ _
+
+#guard
+  (suffixLevenshtein Levenshtein.defaultCost "kitten".toList "sitting".toList).1 =
+    [3, 3, 4, 5, 6, 6, 7]
+
+#guard levenshtein Levenshtein.defaultCost
+  "but our fish said, 'no! no!'".toList
+  "'put me down!' said the fish.".toList = 21
```

Comment:
Should these be in a separate test file? Should we have a style guide that gives an opinion on this?

(no need to un-bors the PR, we can clean this type of thing up another time)